### PR TITLE
Wasm chain integration polish

### PR DESCRIPTION
**Pull request by abizjak**
## Purpose

Addresses #301 

## Changes

- Rename wasm-chain-integration to concordium-smart-contract-engine
- Mainly documentation changes in the code, and some hiding of definitions that should not be public
- factor out the fuzzing into a separate crate wasm-fuzz since the it depends on unpublished libraries.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by limemloh — APPROVED** [review]

File: smart-contracts/README.md [hunk 1]
```diff
@@ -93,7 +93,7 @@ by Nick Fitzerald.
 
 So far the fuzzer discovered three [bugs](wasm-chain-integration/fuzz/fixed_artifacts/interpreter), which we fixed.
 
-## Software requirements
+ ## Software requirements
```
[the remark this answers is not in the document]
Comment (abizjak) [reply smart-contracts/README.md:96]: Indeed.

File: smart-contracts/wasm-chain-integration/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+## concordium-smart-contract-engine
+
+A library that provides an implementation of execution of smart contracts on top
+of the `concordium-wasm` library that implements the underlying Wasm execution.
+
+In particular this library is used by the Concordium node to execute both V0 and
```
[the remark this answers is not in the document]
Comment (abizjak) [reply smart-contracts/wasm-chain-integration/README.md:6]: Added.

File: smart-contracts/wasm-chain-integration/src/lib.rs [hunk 3]
```diff
@@ -1,6 +1,51 @@
+//! This library provides functionality that builds on top of the Wasm engine
```
[the remark this answers is not in the document]
Comment (abizjak) [reply smart-contracts/wasm-chain-integration/src/lib.rs:1]: Added

File: smart-contracts/wasm-chain-integration/src/v1/mod.rs [hunk 4]
```diff
@@ -1355,8 +1365,12 @@ pub enum InvokeFailure {
     SendingV0Failed,
     /// Invoking a contract failed with a runtime error.
     RuntimeError,
+    /// The module to upgrade
```
[the remark this answers is not in the document]
Comment (abizjak) [reply smart-contracts/wasm-chain-integration/src/v1/mod.rs:1368]: Added

File: smart-contracts/wasm-fuzz/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,42 @@
+[package]
+name = "wasm-chain-integration-fuzz"
```
[the remark this answers is not in the document]
Comment (abizjak) [reply smart-contracts/wasm-fuzz/Cargo.toml:2]: It slipped. The name does not matter for anything. But is is changed now.

File: smart-contracts/wasm-chain-integration/src/v0/mod.rs [hunk 6]
```diff
@@ -1102,7 +1176,7 @@ fn reason_from_wasm_error_code(n: i32) -> ExecResult<i32> {
     Ok(n)
 }
 
-/// Invokes an receive-function from a given artifact *bytes*
+/// Invokes a receive-function from a given artifact *bytes*
```
[the remark this answers is not in the document]
Comment (abizjak) [reply smart-contracts/wasm-chain-integration/src/v0/mod.rs:1200]: Yes. I added a section at the beginning of the module pointing to the most important methods, and explanations of the auxiliaries.

**Pull request merged**